arm64: dts: qcom: msm8939-samsung-a7: Add initial dts (v5)

wonderfulShrineMaidenOfParadise opened the pull request:
This dts adds support for Samsung Galaxy A7 smartphone released in 2015.

Add a device tree for A7 with initial support for:

- GPIO keys
- Hall Sensor
- SDHCI (internal and external storage)
- USB Device Mode
- UART (on USB connector via the SM5502 MUIC)
- WCNSS (WiFi/BT)
- Regulators


v2: relocate memory to fit msm8916-samsung-a2015-common.dtsi
v3: Use msm8939-pm8916.dtsi. Set cd-gpios with GPIO_ACTIVE_LOW
v4: Drop unused cont_splash_mem: framebuffer@83000000 and &blsp_dma
[vknecht: adapted for current msm8939.dtsi]
v5: Rename nodes and move to msm8916-mainline
Add suffix "-state" to pinctrl
Drop reserved@86680000 and rmtfs@86700000 which are already included
in msm8939.dtsi
Put status = "okay"; at the end of the properties

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add touch key (v2)

Add the CORERIVER TC360 touch key together with the necessary
fixed regulator for it.

Note that for some reason Samsung decided to connect this to GPIOs
where no hardware I2C bus is available, so we need to fall back
to software bit-banging using i2c-gpio.

v2: Add suffix "-state" to pinctrl.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

Commit:
squash! MSM8916: drm/panel: Generate using linux-mdss-dsi-panel-driver-generator

X-Code-Generator: msm8916-mainline/linux-panel-drivers@999b342
Signed-off-by: lmdpdg <lmdpdg@localhost>

bolilingmeng89006 committed:
drm/panel: s6e3fa2-ams549eh01: Add missing mipi_dsi_dcs_set_display_on()

Without mipi_dsi_dcs_set_display_on(), s6e3fa2-ams549eh01 won't be
enabled on A7.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -0,0 +1,828 @@
+	status = "okay";
+
+	touchscreen@24 {
+		compatible = "cypress,cyttsp5";
```
Waiting for the driver. We can use the old one or the latest one in upstream.

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add accelerometer/magnetometer (v2)

A7 uses a Bosch BMC150 accelerometer/magnetometer combo.
The chip provides two separate I2C devices for the accelerometer
and magnetometer that are already supported by the bmc150-accel
and bmc150-magn driver.

The difference is the way the sensor on A7 is mounted on the mainboard - set
the mount-matrix in the device-specific device tree part to handle that
difference.

v2: Add suffix "-state" to pinctrl.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add rt5033 battery (v2)

The Samsung Galaxy A7 uses a Richtek RT5033 PMIC as battery
fuel gauge, charger, flash LED and for some regulators. For now,
only add the fuel gauge/battery device to the device tree,
so we can check the remaining battery percentage.

The other RT5033 drivers need some more work first before
they can be used properly.

v2: Add suffix "-state" to pinctrl.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add NFC (v2)

The Galaxy A7 has a NXP PN547, which is supported
by the nxp-nci-i2c driver in mainline. It seems to detect NFC tags
using "nfctool" just fine, although more testing is difficult given
there seem to be very few useful applications making use of the
Linux NFC subsystem. :(

Note that for some reason Samsung decided to connect the I2C pins
to GPIOs where no hardware I2C bus is available, so we need to
fall back to software bit-banging with i2c-gpio.

v2: Add suffix "-state" to pinctrl.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add vibrator (v2)

A7 uses motor driver controlled with PWM signal with a fixed regulator
that powers the driver and is controlled by enable signal.

v2: Add suffix "-state" to pinctrl.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add touchscreen (v2)

A7 uses a Cypress TrueTouch Generation 5 that is connected to blsp_i2c5.
Add it to the device tree.

v2: Add suffix "-state" to pinctrl. Use compatible = "cypress,tt21000";
Rename vcpin to vddio

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add display panel (v4)

v2: Set &dsi0 and &dsi0_phy with status = "okay";
v3: adapted for current msm8939.dtsi (mdss_dsi0 and mdss_dsi0_phy)
[vknecht: adapted for current msm8939.dtsi (mdss_dsi0 and mdss_dsi0_phy)]
v4: adapted for current msm8939.dtsi (dsi0 and dsi_phy0)
Add suffix "-state" to pinctrl.
Put status = "okay"; at the end of the properties

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add audio jack detection (v2)

v2: Add suffix "-state" to pinctrl.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

wonderfulShrineMaidenOfParadise commented:
```
[   26.493253] i2c_qup 78b6000.i2c: using default clock-frequency 100000
[   26.493292] i2c_qup 78b6000.i2c: 
                tx channel not available
[   26.494024] i2c_qup 78b5000.i2c: using default clock-frequency 100000
[   26.494054] i2c_qup 78b5000.i2c: 
                tx channel not available
[   26.495417] platform 78b6000.i2c: deferred probe pending
[   26.495446] platform 78d9000.usb: deferred probe pending
[   26.495461] platform 78b5000.i2c: deferred probe pending
[   26.495476] platform 7702000.sound: deferred probe pending
```
i2c is broken.

wonderfulShrineMaidenOfParadise commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -0,0 +1,830 @@
+	};
+};
+
+&blsp_dma {
+	status = "okay";
+};
+
```
It can be dropped with #256 applied. Enable it here for now.

wonderfulShrineMaidenOfParadise commented on drivers/gpu/drm/panel/msm8916-generated/panel-samsung-s6e3fa2-ams549eh01.c:
```diff
@@ -0,0 +1,301 @@
+
+	dsi->lanes = 4;
+	dsi->format = MIPI_DSI_FMT_RGB888;
+	dsi->mode_flags = MIPI_DSI_MODE_VIDEO_BURST;
```
Ideally we could find a way to convert it to video mode. The framerate is at 30 fps in command mode.

wonderfulShrineMaidenOfParadise commented on drivers/gpu/drm/panel/msm8916-generated/panel-samsung-s6e3fa2-ams549eh01.c:
```diff
@@ -0,0 +1,301 @@
+}
+
+static const struct drm_display_mode s6e3fa2_ams549eh01_mode = {
+	.clock = (1080 + 160 + 12 + 36) * (1920 + 12 + 2 + 3) * 60 / 1000,
```
Glitches at full framerate.

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -91,6 +91,64 @@
+			interrupt-parent = <&msmgpio>;
+			interrupts = <20 IRQ_TYPE_EDGE_FALLING>;
+
+			/* vcc/vdd-supply are board-specific */
```
Why are they here then?

wonderfulShrineMaidenOfParadise replied:
https://github.com/torvalds/linux/commit/61f363a625fcbff93171a271b898fcf37dd367c3#diff-532e6f8b9a339cbb86afed3a54807816fd848647116a4cb38f717265856767c1R118
It's copied from a2015.

stephan-gh replied:
They are there on a2015 because the vcc/vdd-supply is not in the common .dtsi but in a3/a5. It doesn't apply here so you should drop this comment.

stephan-gh replied:
The comment is still here? Please remove it :)

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -91,6 +91,64 @@
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+
+		/* NOTE: On some variants e.g. SM-A500FD it's GPIO 91 */
```
A500? I thought this is a7

wonderfulShrineMaidenOfParadise replied:
Uh, typo.

bolilingmeng89006 committed:
drm/panel: panel-samsung-s6e3fa2-ams549eh01: Reduce the framerate to 50 fps

Otherwise it glitches. However, actually it's 25 fps in command mode.

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

stephan-gh commented:
I fixed the default rmtfs in `msm8939.dtsi` in `msm8916/6.1-rc6` so you can drop the `/delete-node/ rmtfs@...` and `reserved@` change after rebasing.

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add initial dts (v5)

This dts adds support for Samsung Galaxy A7 smartphone released in 2015.

Add a device tree for A7 with initial support for:

- GPIO keys
- Hall Sensor
- SDHCI (internal and external storage)
- USB Device Mode
- UART (on USB connector via the SM5502 MUIC)
- WCNSS (WiFi/BT)
- Regulators

v2: relocate memory to fit msm8916-samsung-a2015-common.dtsi
v3: Use msm8939-pm8916.dtsi. Set cd-gpios with GPIO_ACTIVE_LOW
v4: Drop unused cont_splash_mem: framebuffer@83000000 and &blsp_dma
[vknecht: adapted for current msm8939.dtsi]
v5: Rename nodes and move to msm8916-mainline
Add suffix "-state" to pinctrl
Drop reserved@86680000 and rmtfs@86700000 which are already included
in msm8939.dtsi
Put status = "okay"; at the end of the properties

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

bolilingmeng89006 committed:
arm64: dts: qcom: msm8939-samsung-a7: Add sound and modem (v4)

v2: Use msm8916-modem.dtsi
v3: Use msm8939-modem.dtsi
[vknecht: adapted for msm8939-modem.dtsi]
v4: Use msm8916-modem.dtsi

Signed-off-by: Lin, Meng-Bo <linmengbo0689@protonmail.com>

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -78,6 +78,64 @@
+		compatible = "regulator-fixed";
+		regulator-name = "touch_key";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
```
This seems to be 2.8V

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -319,6 +360,52 @@
+		reg = <MI2S_QUATERNARY>;
+		qcom,playback-sd-lines = <1>;
+	};
+};
```
this belongs into the audio commit

stephan-gh commented on arch/arm64/boot/dts/qcom/msm8939-samsung-a7.dts:
```diff
@@ -233,6 +233,32 @@
+		compatible = "regulator-fixed";
+		regulator-name = "tsp_vdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
```
vin-supply?

wonderfulShrineMaidenOfParadise replied:
vin-supply = <&pm8916_s4>;

stephan-gh commented:
Integrated into msm8916/6.1-rc6, thanks!
